db/mysql/client.go

```diff
@@ -0,0 +1,141 @@
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/spf13/viper"
+
+	"github.com/abihaa/hospital-management-system/config"
+	"github.com/abihaa/hospital-management-system/db"
+	domainErr "github.com/abihaa/hospital-management-system/errors"
+	"github.com/abihaa/hospital-management-system/model"
+)
+
+const (
+	patientTableName = "patient"
+)
+
+func init() {
+	db.Register("mysql", NewClient)
+}
+
+//The first implementation.
+type client struct {
+	db *sqlx.DB
+}
+
+func formatDSN() string {
+	cfg := mysql.NewConfig()
+	cfg.Net = "tcp"
+	cfg.Addr = fmt.Sprintf("%s:%s", viper.GetString(config.DbHost), viper.GetString(config.DbPort))
+	cfg.DBName = viper.GetString(config.DbName)
+	cfg.ParseTime = true
+	cfg.User = viper.GetString(config.DbUser)
+	cfg.Passwd = viper.GetString(config.DbPass)
+	return cfg.FormatDSN()
+}
+
+// NewClient initializes a mysql database connection.
+func NewClient(conf db.Option) (db.DataStore, error) {
+	log().Info("initializing mysql connection: " + formatDSN())
+	cli, err := sqlx.Connect("mysql", formatDSN())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to connect to db")
+	}
+	return &client{db: cli}, nil
+}
+
+// To add  patient details.
+func (c *client) SavePatient(patient *model.Patient) (string, error) {
+	if patient.ID != "" {
+		err := c.UpdatePatient(patient.ID, patient)
+		return patient.ID, err
+	}
+
+	patient.ID = uuid.NewV4().String()
+
+	names := patient.Names()
+	if _, err := c.db.NamedExec(fmt.Sprintf(`INSERT INTO %s (%s) VALUES(%s)`, patientTableName, strings.Join(names, ","), strings.Join(mkPlaceHolder(names, ":", func(name, prefix string) string {
+		return prefix + name
+	}), ",")), patient); err != nil {
+		return "", errors.Wrap(err, "failed to add patient details.")
+	}
+
+	return patient.ID, nil
+}
+
+// update patient details.
+func (c *client) UpdatePatient(id string, patient *model.Patient) error {
+	names := patient.Names()
```

Comment:
use patient.Names inline in function call. Dont need to declare variables for things that are just used once.